ibis/expr/analysis.py

```diff
@@ -198,16 +153,20 @@ def finder(expr):
 
 
 def substitute(fn, expr):
-    """
-    Substitute expressions with other expressions.
-    """
-    node = expr.op()
+    """Substitute expressions with other expressions."""
+    try:
+        node = expr.op()
```

Comment:
`expr` should always be an expression type, we should investigate otherwise

Reply:
```python
            arg = tuple(substitute(fn, expr) for expr in arg)
```
doesn't check `arg`'s type which will lead to a recursive call potentially being not-an-`Expr`. I'll fix.

Reply:
Done.